### PR TITLE
Add `use_websocket`

### PR DESCRIPTION
**Pull request by sectore**
Highly inspired by [`use_websocket`](https://github.com/jetli/yew-hooks/blob/main/crates/yew-hooks/src/hooks/use_websocket.rs) of [`yew-hooks`](https://github.com/jetli/yew-hooks)

## Content
- [x] `use_websocket`
- [x] `use_websocket_with_options`
- [x] `auto-reconnect` in case of `Close` or `Error` events.
- [x] Example
- [x]  Book: New chapter `websocket`
- [x] Bump `default-struct-builder`

## Demo
[use_websocket_demo.webm](https://github.com/Synphonyte/leptos-use/assets/47693/ab1ee075-c958-469f-82c1-9ad0d8808109)

### CODE REVIEW

File: src/websocket/use_websocket.rs [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+            UseWebSocketReadyState::Connecting => write!(f, "Connecting"),
+            UseWebSocketReadyState::Open => write!(f, "Open"),
+            UseWebSocketReadyState::Closing => write!(f, "Closing"),
+            UseWebSocketReadyState::Closed => write!(f, "Closed"),
+        }
+    }
+}
+
+/// Options for `WebSocket`.
+// #[derive(DefaultBuilder)]
```
Comment (maccesch) [inline src/websocket/use_websocket.rs:36]: Please also derive `DefaultBuilder` for the options as it makes using options much more economic and is in line with the other functions.

The fields themselves should be made private then.

File: src/websocket/use_websocket.rs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+    /// The current state of the `WebSocket` connection.
+    pub ready_state: ReadSignal<UseWebSocketReadyState>,
+    /// Latest text message received from `WebSocket`.
+    pub message: ReadSignal<Option<String>>,
+    /// Latest binary message received from `WebSocket`.
+    pub message_bytes: ReadSignal<Option<Vec<u8>>>,
+    /// The `WebSocket` instance.
+    pub ws: Option<WebSocket>,
+
+    pub open: OpenFn,
```
Comment (maccesch) [inline src/websocket/use_websocket.rs:94]: Please also document these fields

File: src/websocket/use_websocket.rs [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+    pub close: CloseFn,
+    pub send: SendFn,
+    pub send_bytes: SendBytesFn,
+}
+
+pub fn use_websocket(
```
Comment (maccesch) [inline src/websocket/use_websocket.rs:100]: Please add documentation here. The first line is going to show up in the function overview in the "Functions" chapter in the book. The rest is going to be used by the dedicated page in the book. A link to the demo should be provided which is going to be used to automatically insert the Demo in the book. For the autogeneration to work the function needs to be the first documented item in the file. So please move the structs and enums below.

Please look at any other function to see how it's done there.

If you consider adding more functions in the future (which would be highly appreciated) you can consider using the template as documented in the readme which will take care of a lot of boilerplate.

**Review by maccesch — CHANGES_REQUESTED** [review]
This is awesome, thank you! I found some small stuff...

Would you mind to also add the function to the book? Take a look a the template folder to see where it goes.

**Comment by sectore** [comment]
@maccesch Preview of chapter `use_websocket`. All in 12eb953 , incl. changes to address your review suggestions.

Screenshot 2023-07-11 at 19-23-05 use_websocket - Leptos-Use Guide

File: README.md [hunk 4]
```diff
@@ -57,7 +57,7 @@ cargo test --all-features
 First you need to install
 
 ```shell
-cargo install mdbook-cmdrun trunk
+cargo install mdbook mdbook-cmdrun trunk
```
Comment (sectore) [inline README.md:60]: Quick fix: Installation of `mdbook` crate (not just `mdbook-cmdrun`) is needed, too

File: src/websocket/use_websocket.rs [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+        }
+    }
+}
+
+/// Options for [`use_websocket_with_options`].
+// #[doc(cfg(feature = "websocket"))]
+#[derive(DefaultBuilder, Clone)]
+pub struct UseWebSocketOptions {
+    /// `WebSocket` connect callback.
+    onopen: Option<Box<dyn CloneableFnMutWithArg<Event>>>,
```
Comment (maccesch) [inline src/websocket/use_websocket.rs:378]: I'd like to bug you one final time.
If you [look here](https://github.com/Synphonyte/leptos-use/blob/ef0989cbdd74e4d8180d2654d2206fc55a1384d6/src/use_color_mode.rs#L383C10-L383C10) you can see that we usually omit the `Option` and add a `#[builder(into)] to the `Box<dyn Fn()>` types so the user doesn't need to specify the `Some(...)` or the `Box::new(...)` and can simply specify a plain closure. DX is really close to our hearts 😀

The `Default` impl then just gives an empty closure. This doesn't really affect performance.

**Review by maccesch — CHANGES_REQUESTED** [review]
Thanks for all the quick changes!

File: src/websocket/use_websocket.rs (hunk 5 above)
Comment (sectore) [reply src/websocket/use_websocket.rs]: @maccesch Could you check 2c62c85, pls? I tried to implement your change request, but not sure why `#[builder(into)]`, especially why "simply specify a plain closure" does not work.  It still fails with errors such as:

```bash
error[E0277]: the trait bound `Box<(dyn CloneableFnMutWithArg<Vec<u8>, Output = ()> + 'static)>: From<[closure@use_websocket/src/main.rs:103:30: 103:47]>` is not satisfied
   --> use_websocket/src/main.rs:103:30
    |
103 |               .onmessage_bytes(move |m: Vec<u8>| {
    |  ______________---------------_^
    | |              |
    | |              required by a bound introduced by this call
104 | |                 set_history2.update(|history: &mut Vec<_>| {
105 | |                     history.push(format! {"[onmessage_bytes]: {:?}", m})
106 | |                 });
107 | |             }),
    | |_____________^ the trait `From<[closure@use_websocket/src/main.rs:103:30: 103:47]>` is not implemented for `Box<(dyn CloneableFnMutWithArg<Vec<u8>, Output = ()> + 'static)>`
    |
    = help: the following other types implement trait `From<T>`:
              <Box<OsStr> as From<Cow<'_, OsStr>>>
              <Box<OsStr> as From<OsString>>
              <Box<OsStr> as From<&OsStr>>
              <Box<std::path::Path> as From<Cow<'_, std::path::Path>>>
              <Box<std::path::Path> as From<PathBuf>>
              <Box<std::path::Path> as From<&std::path::Path>>
              <Box<CStr> as From<Cow<'_, CStr>>>
              <Box<CStr> as From<CString>>
            and 19 others
    = note: required for `[closure@use_websocket/src/main.rs:103:30: 103:47]` to implement `Into<Box<(dyn CloneableFnMutWithArg<Vec<u8>, Output = ()> + 'static)>>`
note: required by a bound in `leptos_use::websocket::UseWebSocketOptions::onmessage_bytes`
```

Not sure if there a `From` trait is really needed and how would it be looks like. Any help is appreciated.

Comment (maccesch) [reply src/websocket/use_websocket.rs]: Huh, that's weird indeed! Going to check tonight

Comment (maccesch) [reply src/websocket/use_websocket.rs]: The true error seems to be masked. You can try specifiying the closure with `Box::new(...)` and that might tell you the true error. Just an idea. If it doesn't work I'll investigate further tonight.

Comment (sectore) [reply src/websocket/use_websocket.rs]: > You can try specifiying the closure with Box::new(...) and that might tell you the true error. 

Similar error then:

```bash
error[E0277]: the trait bound `Box<(dyn CloneableFnMutWithArg<web_sys::Event, Output = ()> + 'static)>: From<Box<[closure@use_websocket/src/main.rs:85:30: 85:45]>>` is not satisfied
   --> use_websocket/src/main.rs:85:21
    |
85  |               .onopen(Box::new(move |e: Event| {
    |  ______________------_^
    | |              |
    | |              required by a bound introduced by this call
86  | |                 set_history2.update(|history: &mut Vec<_>| {
87  | |                     history.push(format! {"[onopen]: event {:?}", e.type_()})
88  | |                 });
89  | |             }))
    | |______________^ the trait `From<Box<[closure@use_websocket/src/main.rs:85:30: 85:45]>>` is not implemented for `Box<(dyn CloneableFnMutWithArg<web_sys::Event, Output = ()> + 'static)>`
    |
    = help: the following other types implement trait `From<T>`:
              <Box<OsStr> as From<Cow<'_, OsStr>>>
              <Box<OsStr> as From<OsString>>
              <Box<OsStr> as From<&OsStr>>
              <Box<std::path::Path> as From<Cow<'_, std::path::Path>>>
              <Box<std::path::Path> as From<PathBuf>>
              <Box<std::path::Path> as From<&std::path::Path>>
              <Box<CStr> as From<Cow<'_, CStr>>>
              <Box<CStr> as From<CString>>
            and 19 others
    = note: required for `Box<[closure@use_websocket/src/main.rs:85:30: 85:45]>` to implement `Into<Box<(dyn CloneableFnMutWithArg<web_sys::Event, Output = ()> + 'static)>>`
note: required by a bound in `leptos_use::websocket::UseWebSocketOptions::onopen`
   --> leptos-use/src/websocket/use_websocket.rs:371:10
    |
371 | #[derive(DefaultBuilder)]
    |          ^^^^^^^^^^^^^^ required by this bound in `UseWebSocketOptions::onopen`
...
375 |     onopen: Box<dyn CloneableFnMutWithArg<Event>>,
    |     ------ required by a bound in this associated function
    = note: this error originates in the derive macro `DefaultBuilder` (in Nightly builds, run with -Z macro-backtrace for more info)
```

BTW: Same issue by adding an `on_stop` handler at the [example of `use_scroll_with_options`](https://github.com/Synphonyte/leptos-use/blob/ef0989cbdd74e4d8180d2654d2206fc55a1384d6/examples/use_scroll/src/main.rs#L18-L27)

```bash
error[E0277]: the trait bound `Box<(dyn CloneableFnWithArg<leptos::ev::Event, Output = ()> + 'static)>: From<(Box<[closure@use_scroll/src/main.rs:32:32: 32:35]>,)>` is not satisfied
   --> use_scroll/src/main.rs:32:22
    |
32  |               .on_stop((Box::new(|_| {
    |  ______________-------_^
    | |              |
    | |              required by a bound introduced by this call
33  | |                 log!("scrolling stopped");
34  | |             }),)),
    | |________________^ the trait `From<(Box<[closure@use_scroll/src/main.rs:32:32: 32:35]>,)>` is not implemented for `Box<(dyn CloneableFnWithArg<leptos::ev::Event, Output = ()> + 'static)>`
    |
    = help: the following other types implement trait `From<T>`:
              <Box<OsStr> as From<Cow<'_, OsStr>>>
              <Box<OsStr> as From<OsString>>
              <Box<OsStr> as From<&OsStr>>
              <Box<std::path::Path> as From<Cow<'_, std::path::Path>>>
              <Box<std::path::Path> as From<PathBuf>>
              <Box<std::path::Path> as From<&std::path::Path>>
              <Box<CStr> as From<Cow<'_, CStr>>>
              <Box<CStr> as From<CString>>
            and 19 others
    = note: required for `(Box<[closure@use_scroll/src/main.rs:32:32: 32:35]>,)` to implement `Into<Box<(dyn CloneableFnWithArg<leptos::ev::Event, Output = ()> + 'static)>>`
note: required by a bound in `UseScrollOptions::on_stop`
   --> leptos-use/src/use_scroll.rs:445:10
    |
445 | #[derive(DefaultBuilder)]
    |          ^^^^^^^^^^^^^^ required by this bound in `UseScrollOptions::on_stop`
...
464 |     on_stop: Box<dyn CloneableFnWithArg<web_sys::Event>>,
    |     ------- required by a bound in this associated function
    = note: this error originates in the derive macro `DefaultBuilder` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `use_scroll` (bin "use_scroll") due to previous error

```

Maybe `DefaultBuilder` is missing something?

Comment (maccesch) [reply src/websocket/use_websocket.rs]: Could be! Investigating...

Comment (maccesch) [reply src/websocket/use_websocket.rs]: Ok, this seems to be some closure weirdness. I worked around it by updating the `default-struct-builder` macro to auto un-box stuff. You can now remove the `#[builder(into)]` and it should work when specifying a plain closure.

Comment (sectore) [reply src/websocket/use_websocket.rs]: Nice, thank you! It works great now.

**Review by maccesch — APPROVED** [review]

**Comment by maccesch** [comment]
Hehe the checks are failing because of my changes, not yours… let me fix real quick

**Comment by maccesch** [comment]
Great work, thank you!

**Comment by maccesch** [comment]
To be more in line with VueUse / SolidJS-USE the `use_websocket` function now lives in the root module and is in the "Network" book category. Just FYI.

**Pull request merged**